### PR TITLE
Fix AvoidZeroLengthArrayAllocationsAnalyzer (CA1825) to handle compil…

### LINKED ISSUE #1209: CA1825 reported on compiler generated param array
When attempting to move the analyzers repo to consume our latest analyzers, CA1825 was reported on every single DiagnosticDescriptor object creation that doesn't pass an explicit argument for the param array.

### PR DESCRIPTION
**Pull request by mavasani**
…er generated param array in object creation expression.

The temporary workaround in this analyzer to detect compiler generated param array only handled invocation expressions. This leads to false positives in object creation expressions with param array constructor.

Fixes #1209

### CODE REVIEW

**Comment by mavasani** [comment]
Tagging @dotnet/analyzer-ioperation for review

File: src/Microsoft.NetCore.Analyzers/Core/Runtime/AvoidZeroLengthArrayAllocations.cs [hunk 1]
```diff
@@ -111,13 +111,25 @@ private static bool IsCompilerGeneratedParamsArray(IArrayCreationExpression arra
             var model = context.Compilation.GetSemanticModel(arrayCreationExpression.Syntax.SyntaxTree);
 
             // Compiler generated array creation seems to just use the syntax from the parent.
-            var parent = model.GetOperationInternal(arrayCreationExpression.Syntax, context.CancellationToken) as IInvocationExpression;
-            if (parent?.TargetMethod == null || parent.TargetMethod.Parameters.Length == 0)
+            // Compiler generated array creation seems to just use the syntax from the parent.
```
Comment (jinujoseph) [inline src/Microsoft.NetCore.Analyzers/Core/Runtime/AvoidZeroLengthArrayAllocations.cs]: comment repeated

Comment (mavasani) [inline src/Microsoft.NetCore.Analyzers/Core/Runtime/AvoidZeroLengthArrayAllocations.cs]: thanks, removed.

File: src/Microsoft.NetCore.Analyzers/Core/Runtime/AvoidZeroLengthArrayAllocations.cs [hunk 2]
```diff
@@ -111,13 +111,25 @@ private static bool IsCompilerGeneratedParamsArray(IArrayCreationExpression arra
...
-            if (!parent.TargetMethod.Parameters.Last().IsParams)
+            IMethodSymbol targetMethod;
+            var invocation = parent as IInvocationExpression;
+            if (invocation != null)
+            {
+                targetMethod = invocation.TargetMethod;
+            }
+            else
+            {
+                targetMethod = (parent as IObjectCreationExpression)?.Constructor;
```
Comment (genlu) [inline src/Microsoft.NetCore.Analyzers/Core/Runtime/AvoidZeroLengthArrayAllocations.cs]: `parent` could be `IIndexerAccessExpression` as well.

Comment (mavasani) [inline src/Microsoft.NetCore.Analyzers/Core/Runtime/AvoidZeroLengthArrayAllocations.cs]: Thanks, handled that case and added unit test.

File: src/Microsoft.NetCore.Analyzers/Core/Runtime/AvoidZeroLengthArrayAllocations.cs [hunk 3]
```diff
@@ -111,13 +111,38 @@ private static bool IsCompilerGeneratedParamsArray(IArrayCreationExpression arra
...
-            if (parent?.TargetMethod == null || parent.TargetMethod.Parameters.Length == 0)
+            var parent = model.GetOperationInternal(arrayCreationExpression.Syntax, context.CancellationToken) as IHasArgumentsExpression;
+            if (parent == null)
             {
                 return false;
             }
 
-            if (!parent.TargetMethod.Parameters.Last().IsParams)
+            ISymbol targetSymbol;
+            var invocation = parent as IInvocationExpression;
```
Comment (CyrusNajmabadi) [inline src/Microsoft.NetCore.Analyzers/Core/Runtime/AvoidZeroLengthArrayAllocations.cs]: if (parent is IInvocationExpression invocation)

(same with the rest of your as + null checks).

Comment (mavasani) [inline src/Microsoft.NetCore.Analyzers/Core/Runtime/AvoidZeroLengthArrayAllocations.cs]: Done.

**Pull request merged**